src/MonitorProfiler/ProbeInstrumentation/AssemblyProbePrep.cpp

```diff
@@ -0,0 +1,489 @@
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.SByte"), corLibTypeTokens.systemSByteType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Single"), corLibTypeTokens.systemSingleType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.String"), corLibTypeTokens.systemStringType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UInt16"), corLibTypeTokens.systemUInt16Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UInt32"), corLibTypeTokens.systemUInt32Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UInt64"), corLibTypeTokens.systemUInt64Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.IntPtr"), corLibTypeTokens.systemIntPtrType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UIntPtr"), corLibTypeTokens.systemUIntPtrType);
+
+    return S_OK;
+}
+
+HRESULT AssemblyProbePrep::GetTokenForType(
+    IMetaDataImport* pMetadataImport,
+    IMetaDataEmit* pMetadataEmit,
+    mdToken resolutionScope,
+    tstring name,
+    mdToken& typeToken)
+{
+    IfNullRet(pMetadataImport);
+    IfNullRet(pMetadataEmit);
+
+    HRESULT hr;
+
+    typeToken = mdTokenNil;
+
+    mdTypeRef typeRefToken;
+    hr = pMetadataImport->FindTypeRef(
+        resolutionScope,
+        name.c_str(),
+        &typeRefToken);
+
+    if (FAILED(hr))
+    {
+        IfFailRet(pMetadataEmit->DefineTypeRefByName(
+            resolutionScope,
+            name.c_str(),
+            &typeRefToken));
+    }
+
+    typeToken = typeRefToken;
+    return S_OK;
+}
+
+HRESULT AssemblyProbePrep::GetOrEmitTokenForCorLibAssemblyRef(
+    IMetaDataImport* pMetadataImport,
+    IMetaDataEmit* pMetadataEmit,
+    mdAssemblyRef& corlibAssemblyRef)
+{
+    IfNullRet(pMetadataImport);
+    IfNullRet(pMetadataEmit);
+
+    HRESULT hr;
+    corlibAssemblyRef = mdAssemblyRefNil;
+
+    IfFailRet(HydrateResolvedCorLib());
+
+    ComPtr<IMetaDataAssemblyImport> pMetadataAssemblyImport;
+    IfFailRet(pMetadataImport->QueryInterface(IID_IMetaDataAssemblyImport, reinterpret_cast<void **>(&pMetadataAssemblyImport)));
+
+    // JSFIX: Consider RAII for scope guarding instead of closing this enum in all needed cases.
+    HCORENUM corEnum = 0;
+    mdAssemblyRef mdRefs[ENUM_BUFFER_SIZE];
+    ULONG count = 0;
+
+    const ULONG expectedLength = (ULONG)m_resolvedCorLibName.length();
+    unique_ptr<WCHAR[]> assemblyName(new (nothrow) WCHAR[expectedLength]);
+    IfNullRet(assemblyName);
+
+    while ((hr = pMetadataAssemblyImport->EnumAssemblyRefs(&corEnum, mdRefs, ENUM_BUFFER_SIZE, &count)) == S_OK)
+    {
+        for (ULONG i = 0; i < count; i++)
+        {
+            mdAssemblyRef curRef = mdRefs[i];
+
+            ULONG nameLength = 0;
+            hr = pMetadataAssemblyImport->GetAssemblyRefProps(
+                curRef,
+                nullptr,
+                nullptr,
+                assemblyName.get(),
+                expectedLength,
+                &nameLength,
+                nullptr,
+                nullptr,
+                nullptr,
+                nullptr);
+
+            if (hr == CLDB_S_TRUNCATION)
```

Comment:
The typical approach here is to call GetAssemblyRefProps with just nameLength, and then call it again with a buffer that matches the right size. Alternatively, if you get the error code, reallocate with the right size and try again.